[ENH] Add DeprecationWarning for 3D masker.transform inputs

tsalo opened the pull request:
Closes #2726, closes #3132, and closes #2724.

<!-- Please give a brief overview of what has changed in the PR.
If you're not sure what to write, consider it a note to the maintainers to indicate
what they should be looking for when they review the pull request. -->
Changes proposed in this pull request:

- Fix docstrings of masker classes' `inverse_transform` methods to make it clear that 1D inputs are allowed, and will produce a 3D image.
- Explicitly allow 1D inputs to `NiftiLabelsMasker.inverse_transform`. This is the only masker that did not already accept 1D inputs and produce 3D outputs.
- Raise a DeprecationWarning when providing a 3D image to masker classes' `transform` methods. At the moment, a 3D image will be transformed into a 2D array. Starting in version 0.12, it will be transformed into a 1D array.
- Add tests that cover these input and output shapes, and check for the DeprecationWarning when appropriate.

To do:

- [x] I still need to update `MultiNiftiMasker`.
    - Fortunately, I don't need to. `MultiNiftiMasker` uses the same underlying code for the transform, so it already raises the warning.
- [x] Add whatsnew entry.
- [x] Exhibit expected behavior in examples and documentation.

tsalo commented on nilearn/maskers/base_masker.py:
```diff
@@ -66,6 +66,19 @@ def _filter_and_extract(
             class_name,
             _utils._repr_niimgs(imgs, shorten=False)))
+
+    imgs = _utils.check_niimg(imgs)
+
+    # Raise warning if a 3D niimg is provided.
+    if imgs.ndim == 3:
+        warnings.warn(
+            'Starting in version 0.12, 3D images will be transformed to '
+            '1D arrays. '
+            'Until then, 3D images will be coerced to 2D arrays, with a '
+            'singleton first dimension representing time.',
+            DeprecationWarning,
+        )
+
     imgs = _utils.check_niimg(imgs, atleast_4d=True, ensure_ndim=4,
                               dtype=dtype)
```
`imgs` may be a string (or list of strings, I guess) at this point. I therefore needed to look at the Nifti1Image object first, _before_ changing its dimensionality with `check_niimg(..., ensure_ndim=4)`. The best solution I could come up with was to call `check_niimg` twice.

tsalo commented on nilearn/maskers/nifti_labels_masker.py:
```diff
@@ -633,11 +637,19 @@ def inverse_transform(self, signals):
 
         Any mask given at initialization is taken into account.
 
+        .. versionchanged:: 0.9.2dev
+
+            This method now supports 1D arrays, which will produce 3D images.
```
This should be the only actual behavior change in this PR.

bthirion reviewed (commented):
Maybe we can factor a bit more but overall that sounds right.

tsalo commented on nilearn/maskers/tests/test_nifti_maps_masker.py:
```diff
@@ -5,23 +5,36 @@
-    data = rand_gen.standard_normal(size=(shape + (length,)))
-    return nibabel.Nifti1Image(data, affine), nibabel.Nifti1Image(
-        as_ndarray(data[..., 0] > 0.2, dtype=np.int8), affine)
+def generate_random_img(
```
I can move it into `data_gen`.

tsalo commented on nilearn/maskers/tests/test_nifti_masker.py:
```diff
@@ -404,7 +420,7 @@ def test_filter_and_mask():
 
     masker = NiftiMasker()
-    params = get_params(NiftiMasker, masker)
+    params = class_inspect.get_params(NiftiMasker, masker)
```
I suppose it isn't, since extra stuff will be imported. I've switched that back.

ymzayek reviewed (commented):
@tsalo LGTM! I will take a bit more time to review the tests. Just quickly wondering why you changed the way `Nifti1Image` is imported in the tests. But I guess this is quite inconsistent throughout the library and I believe both ways are equally correct so it's not a big deal.

tsalo commented:
I changed imports in a few ways, so I'm not sure which specific change you're referring to. The main changes I recall making are (1) reordering imports based on `isort` and (2) consolidating redundant imports. For example, in some cases I've seen something like:

```python
import nibabel
...
from nibabel import Nifti1Image  # redundant with the overall nibabel import earlier
```

in which case I removed the redundant import. It could be that there are cases where I made changes that were less clean-cut (e.g., importing a module instead of many individual functions), and if you'd prefer to revert those please just flag them in your review.

ymzayek reviewed (approved):
Nothing else to add, thx @tsalo !

NicolasGensollen reviewed (commented):
Thanks @tsalo !
LGTM, just a small documentation suggestion.

tsalo commented on nilearn/regions/signal_extraction.py:
```diff
@@ -202,10 +202,12 @@ def signals_to_img_labels(signals, labels_img, mask_img=None,
-    ##     data[labels_data == label, :] = signals[:, n]
-    data = np.zeros(target_shape + (signals.shape[0],),
-                    dtype=signals.dtype, order=order)
+    # for n, label in enumerate(labels):
```
Thanks! Should be addressed in https://github.com/nilearn/nilearn/pull/3322/commits/9522593d97a6aca8b53acdeb21c0bd86f1aef5b4.

NicolasGensollen reviewed (approved):
LGTM, thanks @tsalo !

ymzayek commented:
Can we merge this one? The failure is unrelated and is now fixed

tsalo commented:
There are two approving reviews, so it should be okay to merge. @bthirion do you want to take another look before we merge?

bthirion commented on nilearn/maskers/tests/test_nifti_masker.py:
```diff
@@ -219,29 +218,31 @@ def test_4d_single_scan():
 
 def test_5d():
-    mask = np.zeros((10, 10, 10))
+    """Test that list of 4D images with last dim=1 are treated as 3d."""
```
I think that this docstring is not accurate.

tsalo replied:
Thanks for catching that. It should be fixed now.

bthirion reviewed (commented):
There seems to be a minor glitch otherwise the PR LGTM.

bthirion reviewed (approved):
All right, thx !

ymzayek commented:
Thx all! Merging
